### PR TITLE
Update registry defaulting behavior

### PR DESCRIPTION
**Pull request by ResidentMario**
A recent PR added a concept of a default registry to user configuration, allowing users to under-specify the parameters in `push` and use the API in a more convenient way.

This PR tweaks how `push` interacts with the default remote registry, using it to parameterize the `dest` kwarg instead of the `registry` kwarg. This is how I expected the API to work, and it makes more sense to me.

Old behavior:

```python
# assume that the default registry is 's3://my-bucket'
p = t4.Package()

# == p.push('foo/bar', dest='s3://other-bucket/foo/bar', registry='s3://my-bucket/')
p.push("foo/bar", "s3://other-bucket")
```

This is confusing because it looks like I'm pushing `foo/bar` to the `other-bucket` S3 bucket. However, instead I am pushing the _files_ to `s3://other-bucket`, and the _package manifest_ to `s3://my-bucket`. From my perspective as an end user when I go to visit `s3://other-bucket` it looks like my package disappeared!

New behavior:

```python
# assume that the default registry is 's3://my-bucket'
p = t4.Package()

# == p.push('foo/bar', dest='s3://my-bucket/foo/bar', registry='s3://my-bucket/')
p.push("foo/bar")   

# == p.push('foo/bar', dest='s3://other-bucket/foo/bar', registry='s3://other-bucket/')
p.push("foo/bar", dest="s3://other-bucket/foo/bar")

# raises; 'dest' is still a required argument
p.push("foo/bar", registry="s3://my-bucket/")
```

Changes:
* The default registry is now only used if only the package name is specified (e.g. `p.push('foo/bar')`).
* `registry` is inferred from `dest`. If the target is `s3://` the root of the bucket is passed as the registry, if the target is `file:///` the path is used as the registry.

If you want to deposit the files in one bucket and the manifest in another you can still do that with a fully specified `push`.

### CODE REVIEW

File: api/python/t4/packages.py [hunk 1]
```diff
@@ -859,13 +859,35 @@ def push(self, name, dest, registry=None, message=None):
         validate_package_name(name)
         self._set_commit_message(message)
 
-        if registry is None:
+        # if only a package name is set, set registry and dest to the default registry
```
Comment (stevededalus) [inline api/python/t4/packages.py]: Make these comments sentences (Capital letter + punctuation)

File: api/python/t4/packages.py [hunk 2]
```diff
@@ -859,13 +859,35 @@ def push(self, name, dest, registry=None, message=None):
...
+            if parsed.scheme == 's3':
+                bucket, _, _ = parse_s3_url(parsed)
+                registry = 's3://' + bucket
+            elif parsed.scheme == 'file':
+                registry = parsed.path
+            else:
+                raise NotImplementedError
+
+        # specifying registry but not dest doesn't make sense
+        elif dest is None and registry is not None:
```
Comment (stevededalus) [inline api/python/t4/packages.py]: this feels like a case where two nested if statements are more readable 

```
if dest is none:
    if registry is none:
...
```

File: api/python/tests/integration/test_packages.py [hunk 3]
```diff
@@ -111,12 +111,13 @@ def test_default_registry(tmpdir):
...
             assert test_file.resolve().as_uri() == pkg['bar'].physical_keys[0]
 
+    # TODO: fix this test?
     with patch('t4.packages.get_remote_registry') as mock_config:
         mock_config.return_value = new_base_path
         new_pkg.push("Quilt/Test", Path(tmpdir, 'test_dest').resolve().as_uri())
         with open(out_path) as fd:
             pkg = Package.load(fd)
-            assert pkg['bar'].physical_keys[0].endswith('test_dest/Quilt/Test/bar')
+            assert pkg['bar'].physical_keys[0].endswith('test_default_registry0/bar')
```
Comment (stevededalus) [inline api/python/tests/integration/test_packages.py]: what's going on with this `default_registry0` string?

File: api/python/tests/integration/test_packages.py [hunk 4]
```diff
@@ -260,6 +261,7 @@ def test_load_into_t4(tmpdir):
 
         # Manifest copied
         top_hash = new_pkg.top_hash()
+        # FIXME: understand what is happening with these byte calls
```
Comment (stevededalus) [inline api/python/tests/integration/test_packages.py]: What's this comment mean?

Comment (ResidentMario) [reply api/python/tests/integration/test_packages.py]: Old debug line, removed.

File: api/python/tests/integration/test_packages.py (hunk 3 above)
Comment (ResidentMario) [reply api/python/tests/integration/test_packages.py]: Updated the test, it now actually exercises the `get_remote_registry` code path.

**Pull request merged**